model-datastructure/src/commonMain/kotlin/org/modelix/model/operations/BulkUpdateOp.kt

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.modelix.model.operations
+
+import org.modelix.model.api.ITree
+import org.modelix.model.api.IWriteTransaction
+import org.modelix.model.lazy.CLTree
+import org.modelix.model.lazy.IDeserializingKeyValueStore
+import org.modelix.model.lazy.KVEntryReference
+import org.modelix.model.persistent.CPTree
+import org.modelix.model.persistent.IKVValue
+import org.modelix.model.persistent.SerializationUtil
+
+class BulkUpdateOp(
+    val resultTreeHash: KVEntryReference<CPTree>,
+    val subtreeRootId: Long,
+) : AbstractOperation() {
+
+    override fun getReferencedEntries(): List<KVEntryReference<IKVValue>> = listOf(resultTreeHash)
+
+    /**
+     * Since this operation is recorded at the end of a bulk update we need to create an IAppliedOperation without
+     * actually applying it again.
+     */
+    fun afterApply(baseTree: CLTree) = Applied(baseTree)
+
+    override fun apply(transaction: IWriteTransaction, store: IDeserializingKeyValueStore): IAppliedOperation {
+        val baseTree = transaction.tree as CLTree
+        val resultTree = getResultTree(store)
+        TODO("Change the (sub)tree so that it is identical to the resultTree")
+        return Applied(baseTree)
+    }
+
+    private fun getResultTree(store: IDeserializingKeyValueStore): CLTree = CLTree(resultTreeHash.getValue(store), store)
+
+    override fun toString(): String {
+        return "BulkUpdateOp ${resultTreeHash.getHash()}, ${SerializationUtil.longToHex(subtreeRootId)}"
+    }
+
+    override fun toCode(): String {
+        return """// TODO BulkUpdateOp"""
```

Comment:
Follow-up ticket to reference here?

Reply:
I just removed `IOperation.toCode()`. It's dead code that was used previously to generate the test cases in [ConflictResolutionTest](https://github.com/modelix/modelix.core/blob/main/model-datastructure/src/commonTest/kotlin/ConflictResolutionTest.kt).